Fix native BlockAlignmentControl

dcalhoun opened the pull request:
## Related PRs

* https://github.com/wordpress-mobile/gutenberg-mobile/pull/4042

## Description
Fixes #35173. Relates to #34710. A recent web-centric change introduced usage of `MenuGroup` and `MenuItem`, both of which are unsupported in the native editor currently. This updates the conditional within `BlockAlignmentControl` to reinstate the previous UI for the native platform.

## How has this been tested?

### Native Alignment

1. Add a Cover block within the editor. 
2. Press the Alignment icon within the block toolbar. 
3. Change the selected alignment from None to another option.
4. **Expected:** Verify the correct alignment is persisted for the block. 

### Web Alignment
Complete Testing Instructions found in https://github.com/WordPress/gutenberg/pull/34710. 

## Screenshots <!-- if applicable -->
n/a

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dcalhoun committed:
Fix native BlockAlignmentControl

A recent web-centric change introduced usage of `MenuGroup` and
`MenuItem`, both of which are unsupported in the native editor
currently. This updates the conditional within `BlockAlignmentControl`
to reinstate the previous UI for the native platform.

fluiddot commented on packages/block-editor/src/components/block-alignment-control/test/index.native.js:
```diff
@@ -0,0 +1,37 @@
+	);
+	const alignButton = screen.getByA11yLabel( 'Align' );
+	fireEvent.press( alignButton );
+	const noneAlignmentButton = screen.getByA11yLabel( 'Align center' );
+	fireEvent.press( noneAlignmentButton );
```
Since we're getting here the align center button, what do you think about renaming this to `centerAlignmentButton`?

dcalhoun replied:
Definitely. Thank you for catching this mistake. Addressed in https://github.com/WordPress/gutenberg/pull/35191/commits/bcd16d620af229b66d2f3dd3b22f293552005144.

fluiddot commented on packages/block-editor/src/components/block-alignment-control/test/index.native.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import BlockAlignmentUI from '../ui';
+
+it( 'should call onChange with undefined, when the control is already active', () => {
```
Regarding the `when the control is already active` part, do we want to test that `onChange` is called with `undefined` when tapping the currently active alignment value? If this is the case, on line `18` we should get the left alignment button, right?

dcalhoun replied:
Great catch. The odd pairing of relying upon the default left alignment and pressing "None", meant this test passed even though it was not truly the "selected" option. I addressed this in https://github.com/WordPress/gutenberg/pull/35191/commits/bcd16d620af229b66d2f3dd3b22f293552005144 by shifting the test right alignment to be more explicit.

fluiddot replied:
Thanks for addressing it ❤️ !

I'm thinking that the previous code was covering the new logic of the `None` option. Maybe it's worth it to include that test case (having a selected alignment option and select `None`), wdyt?

fluiddot commented on test/native/__mocks__/styleMock.js:
```diff
@@ -135,4 +135,7 @@ module.exports = {
 	buttonNoBg: {
 		color: 'orange',
 	},
+	isSelected: {
+		color: 'blue',
+	},
```
Mocking this style is required due to rendering `BlockAlignmentUI` in the `block-alignment-control` test, right?

dcalhoun replied:
Correct. Further down the tree the following code is encountered, which requires a mock of the style attribute referenced directly. https://github.com/WordPress/gutenberg/blob/e5cb098f718d7d1ee1fc2946374c9388e1f0cb2e/packages/components/src/mobile/bottom-sheet/cell.native.js#L412

fluiddot replied:
Thanks for clarifying this ❤️ !

fluiddot reviewed (commented):
I verified that the issue is no longer happening with these changes 🎊 , awesome job @dcalhoun 💯 .

The only thing I'd like to double-check before approving the PR is if the `none` alignment value is being properly handled in the Image block. I noticed that the none/default alignment value for that block aligns the content to the left (see attached video):

https://user-images.githubusercontent.com/14905380/135136639-cba0a05a-49a7-4653-86fe-ed15353ce71a.mp4

**Tested on Simulator iPhone 12 Pro Max (iOS 14.5).**

dcalhoun committed:
Fix nuanced test irregularities

Rename mistakenly named assignments, which were misleading.
Additionally, the reliance upon the default `value="left"` could lead to
a false-positive test. Changing to a none-default of `value="right"`
improves the validity of the test.

dcalhoun commented:
> The only thing I'd like to double-check before approving the PR is if the `none` alignment value is being properly handled in the Image block. I noticed that the none/default alignment value for that block aligns the content to the left 

@fluiddot this is a fantastic question. Thanks for bringing this up. 

In regards to explicit code to "handle" a `none` alignment, currently it does not exist. https://github.com/WordPress/gutenberg/blob/bcd16d620af229b66d2f3dd3b22f293552005144/packages/block-library/src/image/edit.native.js#L623-L629

However, I would pose the question: _what should the representation be for the `none` alignment?_ 

From reviewing the web representation, it might imply that representing `none` with a left-aligned image in the editor would be expected, i.e. what we render currently in your example video. The following showcases the `none` and `left` alignment options in both the web editor and web preview. 

<details><summary>Web Editor (top) vs Web Preview (bottom)</summary>

| Alignment: None | Alignment: Left |
| - | - |
| ![web-editor-none](https://user-images.githubusercontent.com/438664/135155721-ff4837f6-26a3-476d-a2fc-cdca35b92829.png) | ![web-editor-left](https://user-images.githubusercontent.com/438664/135156140-4ec0337a-d33a-4583-a1ad-122abd5eba27.png) |
| ![web-preview-none](https://user-images.githubusercontent.com/438664/135155776-35a7ae4a-5671-4260-bc3c-44b4c9da428f.png) | ![web-preview-left](https://user-images.githubusercontent.com/438664/135155823-bbe446d1-e006-4e28-bdd8-8171653e3095.png) |

</details>

Given that, I would say our two options are presenting `none` alignments either centered- or left-aligned. Currently, I think the latter may make the most sense and aligns with the web editor representation. WDYT? 

<details><summary>Native Editor `none` Presentation</summary>

| Center | Left |
| - | - |
| ![none-option-center](https://user-images.githubusercontent.com/438664/135156015-688022bd-a1ff-47c4-a36d-dc97f62245f1.png) | ![none-option-left](https://user-images.githubusercontent.com/438664/135156035-1e51d757-d315-42f8-937a-6596375c604a.png) |

</details>

ntsekouras reviewed (approved):
Thanks for handling this @dcalhoun ! 💯 

I can see that the main change is `extraProps = isToolbar || Platform.isNative...` so the web version is not affected at all.

Just to learn a bit more, can you please share why MenuGroup and MenuItem are not supported in native? Is there any technical limitation with native?

fluiddot commented:
> @fluiddot this is a fantastic question. Thanks for bringing this up.
> 
> In regards to explicit code to "handle" a `none` alignment, currently it does not exist.

Thanks for checking this ❤️ !

> However, I would pose the question: _what should the representation be for the `none` alignment?_
> 
> From reviewing the web representation, it might imply that representing `none` with a left-aligned image in the editor would be expected, i.e. what we render currently in your example video. The following showcases the `none` and `left` alignment options in both the web editor and web preview.
>
> Given that, I would say our two options are presenting `none` alignments either centered- or left-aligned. Currently, I think the latter may make the most sense and aligns with the web editor representation. WDYT?

Great investigation 💯 ! I agree that only center and left alignments fit what I would expect when having `none` alignment value. For the Image block, the current approach with content left-aligned makes sense to me as it matches the same behavior of the web version. It's interesting that this block is the only one that handles the `none` case this way, I noticed that the rest (Embed, Cover, etc.) by default display the content centered.

As a side note, while testing I spotted that the theme might affect the Image block as well, as an example, I used the theme `Spearhead` and the `none` value on the preview is centering the image 🤔 .

Web editor|Web Preview
--|--
<img width="1020" alt="Screenshot 2021-09-29 at 10 44 01" src="https://user-images.githubusercontent.com/14905380/135234478-501c7aab-7426-49de-b887-6ac328d1e67b.png">|<img width="997" alt="Screenshot 2021-09-29 at 10 42 49" src="https://user-images.githubusercontent.com/14905380/135234485-ca99ce55-5595-425e-b121-a6b095d07218.png">

fluiddot commented on packages/block-editor/src/components/block-alignment-control/test/index.native.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import BlockAlignmentUI from '../ui';
+
+it( 'should call onChange with undefined, when the control is already active', () => {
```
I removed the comma to match the same format as the following test case's description.
```suggestion
it( 'should call onChange with undefined when the control is already active', () => {
```

fluiddot reviewed (approved):
LGTM 🎊 !

_Tested `Native Alignment` and `Web Alignment` test cases._

mtias commented:
Also curious about `MenuGroup` and any timelines for supporting it.

dcalhoun committed:
Fix inconsistent test description

Co-authored-by: Carlos Garcia <fluiddot@gmail.com>

dcalhoun commented:
> Just to learn a bit more, can you please share why MenuGroup and MenuItem are not supported in native? Is there any technical limitation with native?

> Also curious about `MenuGroup` and any timelines for supporting it.

@ntsekouras @mtias in regards to lack of `MenuGroup` support, there is no technical limitation but likely originates from no specific mobile-centric UI or internal priority requiring it. My general understanding of the focus of internal mobile teams currently largely falls into two categories: increased block support and site/page editing features. So, I would imagine `MenuGroup` has just not been a necessity for those two categories so far. 

Additionally, the mobile UI often leverages more platform-specific UI like our [`Cell` components](https://github.com/WordPress/gutenberg/tree/dad3eefb4fe53cf4fa0dc0729dc5e311a3b84b56/packages/components/src/mobile/bottom-sheet) or native action sheets. There could be opportunity to encapsulate this platform-specific UI as mobile's version of `MenuGroup`/`MenuItem`, but maybe not. It could be the API is so divergent it does not make sense. 

<details><summary>Cell and Action Sheets</summary>

<img width="250" alt="mobile-cell" src="https://user-images.githubusercontent.com/438664/135279298-ad1f9bca-10e3-4abf-bf36-e89567e97659.jpeg" />
<img width="250" alt="mobile-cell-2" src="https://user-images.githubusercontent.com/438664/135279312-93a99077-65cc-4fe4-8fd8-d961de660ba1.jpeg" />
<img width="250" alt="mobile-action-sheet" src="https://user-images.githubusercontent.com/438664/135278946-93a1d122-b4dc-4160-a4f0-225bf33c5d04.jpeg" />

</details>

In regards to a timeline for support, I would defer to @hypest, @mchowning, @antonis, and @iamthomasbishop to ask if they have any knowledge of plans to leverage a [`MenuGroup` UI](https://github.com/WordPress/gutenberg/tree/trunk/packages/components/src/menu-group). 

I interpret the inquiries regarding `MenuGroup` to be non-blocking, so I will move to merge this regression fix to avoid issues in our next mobile release. Thanks. 🙇🏻

mtias commented:
It's not blocking now, but I want to stay as far away as we can from these platform divergences on shared components, because we are shipping extra code for anyone and just makes maintenance harder.

hypest commented:
👋 thanks for the ping @dcalhoun!

There are no current plans for supporting `MenuGroup`s, but now it's actually a good time to look into it. David has already pinged you above @iamthomasbishop but yeah, please collaborate with @dcalhoun to have a first assessment of such a project. Thanks!

> I want to stay as far away as we can from these platform divergences on shared components, because we are shipping extra code for anyone and just makes maintenance harder.

I agree with @mtias , it's unfortunate to have to ship and maintain platform specific "forks", especially when they end up being divergent designs. I also understand how they come to be some times: a UI change on the web is merged without being tested on native mobile (at least for crashes) which puts the native mobile team in panic mode to fix the crash in any fast/dirty way they can. It'd be better if we get to collaborate before such merges happen, to at least test out native mobile when new UI is introduced (the "none" alignment option in this case).

mtias commented:
@hypest I think we need more participation on the UI components side of things as well, since these are primitives that people making updates to blocks should not have to account for because they should have a representation cross-platform.
